### PR TITLE
feat(risedev): support docker-backed mysql service

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support MySQL service in RiseDev through Docker.

- Internally use `docker run -rm` to start the service.
- May persist data under `PREFIX_DATA` by mapping the volume into the container.
- Stop the docker containers on `risedev k`.

~~Can be further extracted to a common module for launching any external services with Docker.~~

There's a general `DockerService` to launch any external services with Docker. `MySqlService` is simply an alias to `DockerService<MySqlConfig>`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: Makefile.toml [hunk 1]
```diff
@@ -727,6 +727,10 @@ ${TMUX} list-windows -F "#{window_name} #{pane_id}" \
 | awk '{ print $2 }' \
 | xargs -I {} ${TMUX} send-keys -t {} C-c C-d
 
+# Kill docker components
+docker stop $(docker ps -a -q -f name=risedev-) || true
```
Comment (kwannoel) [inline Makefile.toml:731]:
```suggestion
echo "kill docker components"
docker stop $(docker ps -a -q -f name=risedev-) || true
```

File: src/risedevtool/src/task/mysql_service.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+use super::{ExecuteContext, Task};
+use crate::{DummyService, MySqlConfig};
+
+// TODO: extract to a common docker-backed service
+pub struct MysqlService {
+    config: MySqlConfig,
+}
+
+const IMAGE_TAG: &str = "mysql:8";
```
Comment (StrikeW) [inline src/risedevtool/src/task/mysql_service.rs:29]: Better to allow specify the image tag in `risedev.yml`

File: src/risedevtool/src/task/mysql_service.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        if self.config.persist_data {
+            let path = Path::new(&env::var("PREFIX_DATA")?).join(self.id());
+            fs_err::create_dir_all(&path)?;
+            run_cmd
+                .arg("-v")
+                .arg(format!("{}:/var/lib/mysql", path.to_string_lossy()));
+        }
+        run_cmd.arg(IMAGE_TAG);
+
+        ctx.run_command(ctx.tmux_run(run_cmd)?)?;
```
Comment (kwannoel) [inline src/risedevtool/src/task/mysql_service.rs:97]: Why need tmux to run it

File: risedev.yml [hunk 4]
```diff
@@ -1431,20 +1431,28 @@ template:
     # address of redis
     address: "127.0.0.1"
 
-  # MySQL service. Currently user-managed only
+  # MySQL service backed by docker.
```
[the remark this answers is not in the document]
Comment (stdrc) [reply risedev.yml:1434]: +1 for mysql image version

File: src/risedevtool/src/task/mysql_service.rs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+        let mut run_cmd = self.docker_run_args();
+        if self.config.persist_data {
+            let path = Path::new(&env::var("PREFIX_DATA")?).join(self.id());
+            fs_err::create_dir_all(&path)?;
+            run_cmd
+                .arg("-v")
+                .arg(format!("{}:/var/lib/mysql", path.to_string_lossy()));
+        }
+        run_cmd.arg(IMAGE_TAG);
```
Comment (stdrc) [inline src/risedevtool/src/task/mysql_service.rs:95]: Is this cleaned by `risedev clean-data`?

**Review by StrikeW — COMMENTED** [review]
Is it a prerequisite work to support the [inline style e2e test](https://github.com/risingwavelabs/risingwave/issues/12451)?

File: risedev.yml [hunk 6]
```diff
@@ -1431,20 +1431,28 @@ template:
...
   mysql:
     # Id to be picked-up by services
-    id: mysql
+    id: mysql-${port}
```
Comment (stdrc) [inline risedev.yml:1437]: How is this `port` determined?

**Review by xxchan — APPROVED** [review]
Gnerally LGTM. Would like to invite @StrikeW to try this (and also risedev slt

File: src/risedevtool/src/task/mysql_service.rs (hunk 5 above)
Comment (BugenZhao) [reply src/risedevtool/src/task/mysql_service.rs:95]: Yes.

https://github.com/risingwavelabs/risingwave/blob/60b7a82b6ad5f952f583c56cee989d8c009cf502/Makefile.toml#L120-L125

File: risedev.yml (hunk 6 above)
Comment (BugenZhao) [reply risedev.yml:1437]: It'll be expanded to the `port` field. May refer to other components.

https://github.com/risingwavelabs/risingwave/blob/60b7a82b6ad5f952f583c56cee989d8c009cf502/risedev.yml#L1105-L1106

Will be expanded with `DollarExpander`.

https://github.com/risingwavelabs/risingwave/blob/60b7a82b6ad5f952f583c56cee989d8c009cf502/src/risedevtool/src/config/dollar_expander.rs#L21-L25

File: src/risedevtool/src/task/mysql_service.rs (hunk 3 above)
Comment (BugenZhao) [reply src/risedevtool/src/task/mysql_service.rs:97]: No strong reasons but it's just to make it consistent with other components. For example,

- Developers can attach to the session to find MySQL as a window there, just like how other components work.
- The logs are dumped under `.risingwave/log` as well.
- It follows the lifecycle of a `risedev-dev` cluster.

File: risedev.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (BugenZhao) [reply risedev.yml:1434]: Done

**Comment by BugenZhao** [comment]
> ~~Can be further extracted to a common module for launching any external services with Docker.~~
>
> There's a general `DockerService` to launch any external services with Docker. `MySqlService` is simply an alias to `DockerService<MySqlConfig>`.

Extracted right in this PR because I don't find splitting the task into other PRs helpful for reviewing.

**Review by StrikeW — APPROVED** [review]

**Pull request merged**